### PR TITLE
The Frogger-like game implementation

### PR DESCRIPTION
**Pull request by lalalilulu**
### Frogger game

[Demo](https://lalalilulu.github.io/frontend-nanodegree-arcade-game/) |
[Code base](https://github.com/lalalilulu/frontend-nanodegree-arcade-game)

The code is submitted in a dedicated feature branch.
Please, review.

### CODE REVIEW

File: submissions/lalalilulu/frogger-game/app.js [hunk 1]
```diff
@@ -0,0 +1,112 @@
+const dx = 101;
+const dy = 80;
+const collision_offset = 60;
+const player_start_x = dx * 2;
+const player_start_y = dy * 5;
+
+function calculateRandomSpeed() {
+    return Math.random() * (300 - 50) + 50;
```
Comment (OleksiyRudenko) [inline submissions/lalalilulu/frogger-game/app.js:8]: Mysterious numbers. There must be some semantic behind explaining the values.
A case for defining constants.

File: submissions/lalalilulu/frogger-game/app.js [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+
+    render() {
+        ctx.drawImage(Resources.get(this.image), this.x, this.y);
+    }
+}
+
+
+class Enemy extends Character {
+
+    constructor(x, y, player, speed, image = 'images/enemy-bug.png') {
```
Comment (OleksiyRudenko) [inline submissions/lalalilulu/frogger-game/app.js:27]: IDE must have warned you that the `speed` argument is not used.

File: submissions/lalalilulu/frogger-game/app.js [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+                    this.x += dx;
+                }
+                break;
+            case 'up':
+                if (this.y > 0) {
+                    this.y -= dy;
+                }
+                break;
+            case 'down':
+                if (this.y + dy * 2 < ctx.canvas.height - 46) {
```
Comment (OleksiyRudenko) [inline submissions/lalalilulu/frogger-game/app.js:92]: Mysterious number 46. There must be some semantic behind explaining the value.
A case for defining a constant.

File: submissions/lalalilulu/frogger-game/app.js [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+                    this.y += dy;
+                }
+                break;
+        }
+    }
+}
+
+const player = new Player();
+const allEnemies = Array.from(Array(3).keys()).map(y => new Enemy(0, (y * dy) + 60, player));
```
Comment (OleksiyRudenko) [inline submissions/lalalilulu/frogger-game/app.js:101]: Mysterious numbers 3 and 60. There must be some semantic behind explaining the values.
A case for defining constants.

`Array.from(Array(3).keys())` is an anti-pattern. What if we have 1000 enemies?
Don't overexploit Array methods where it is not necessary. `for` is a nice iteration method and still worth of being used.

If you had an array of rows to launch enemies at e.g. `[0, 1, 2, 0]` for 4 enemies then `.map` would be an appropriate approach here.

**Review by OleksiyRudenko — CHANGES_REQUESTED** [review]
@lalalilulu a really god job!
Let's have it polished.

File: submissions/lalalilulu/frogger-game/app.js (hunk 1 above)
Comment (lalalilulu) [reply submissions/lalalilulu/frogger-game/app.js]: These mysterious numbers for minimum and maximum enemy speed. I have created the constants for values

File: submissions/lalalilulu/frogger-game/app.js (hunk 2 above)
Comment (lalalilulu) [reply submissions/lalalilulu/frogger-game/app.js]: I have removed the speed variable from the constructor

File: submissions/lalalilulu/frogger-game/app.js (hunk 3 above)
Comment (lalalilulu) [reply submissions/lalalilulu/frogger-game/app.js]: the constant created

File: submissions/lalalilulu/frogger-game/app.js [hunk 5]
```diff
@@ -0,0 +1,117 @@
...
+                    this.y += dy;
+                }
+                break;
+        }
+    }
+}
+
+const player = new Player();
+const arrayOfRows = [0,1,2];
+const allEnemies = arrayOfRows.map(y => new Enemy(0, (y * dy) + icon_height, player));
```
Comment (OleksiyRudenko) [inline submissions/lalalilulu/frogger-game/app.js:106]: This approach is still questionable.
I'd consider either of the below
```
const allEnemies = [];

// Strategy A: X enemies per row strategy
for (let row = 0; row < ENEMY_ROWS; row++) {
  for (let enemyCount = 0; enemyCount < ENEMIES_PER_ROW; enemyCount++) {
    allEnemies.push(new Enemy(0, row * dy + icon_height, player));
  }
}

// Strategy B: X enemies spread across all rows
for (let enemyCount = 0; enemyCount < ENEMIES_PER_ROW; enemyCount++) {
  allEnemies.push(new Enemy(0, (enemyCount % ENEMY_ROWS) * dy + icon_height, player));
}
```

**Review by OleksiyRudenko — APPROVED** [review]
@lalalilulu great job done!
Check the comment below to consider alternative strategies.
Although your current approach would be really OK if you had to customize per enemy anything else but rows they are to appear on.

**Pull request merged**